File: bindings/js/packages/clasp-core/src/builder.ts
/**
 * CLASP client builder
 */

import { Clasp } from './client';
import { ConnectOptions } from './types';

/**
 * Builder for CLASP client
 */
export class ClaspBuilder {
  private _url: string;
  private _options: ConnectOptions = {};

  constructor(url: string) {
    this._url = url;
  }

  /**
   * Get the configured URL
   */
  getUrl(): string {
    return this._url;
  }

  /**
   * Get the configured name
   */
  getName(): string | undefined {
    return this._options.name;
  }

  /**
   * Get the configured features
   */
  getFeatures(): string[] | undefined {
    return this._options.features;
  }

  /**
   * Get the configured token
   */
  getToken(): string | undefined {
    return this._options.token;
  }

  /**
   * Get the reconnect setting
   */
  getReconnect(): boolean | undefined {
    return this._options.reconnect;
  }

  /**
   * Get the reconnect interval
   */
  getReconnectInterval(): number | undefined {
    return this._options.reconnectInterval;
  }

  /**
   * Get all options (for testing)
   */
  getOptions(): ConnectOptions {
    return { ...this._options };
  }

  /**
   * Set client name
   */
  name(name: string): this {
    this._options.name = name;
    return this;
  }

  /**
   * Set client name (alias)
   */
  withName(name: string): this {
    return this.name(name);
  }

  /**
   * Set supported features
   */
  features(features: string[]): this {
    this._options.features = features;
    return this;
  }

  /**
   * Set supported features (alias)
   */
  withFeatures(features: string[]): this {
    return this.features(features);
  }

  /**
   * Set authentication token
   */
  token(token: string): this {
    this._options.token = token;
    return this;
  }

  /**
   * Set authentication token (alias)
   */
  withToken(token: string): this {
    return this.token(token);
  }

  /**
   * Enable/disable auto-reconnect
   */
  reconnect(enabled: boolean): this {
    this._options.reconnect = enabled;
    return this;
  }

  /**
   * Enable/disable auto-reconnect with optional interval
   */
  withReconnect(enabled: boolean, intervalMs?: number): this {
    this._options.reconnect = enabled;
    if (intervalMs !== undefined) {
      this._options.reconnectInterval = intervalMs;
    }
    return this;
  }

  /**
   * Set reconnect interval in milliseconds
   */
  reconnectInterval(ms: number): this {
    this._options.reconnectInterval = ms;
    return this;
  }

  /**
   * Build and connect
   */
  async connect(): Promise<Clasp> {
    const client = new Clasp(this._url, this._options);
    await client.connect();
    return client;
  }
}

File: crates/clasp-client/src/builder.rs
//! Client builder pattern

use crate::{Clasp, Result};

/// Builder for Clasp client
pub struct ClaspBuilder {
    url: String,
    name: String,
    features: Vec<String>,
    token: Option<String>,
    reconnect: bool,
    reconnect_interval_ms: u64,
    #[cfg(feature = "p2p")]
    p2p_config: Option<clasp_core::P2PConfig>,
}

impl ClaspBuilder {
    /// Create a new builder
    pub fn new(url: &str) -> Self {
        Self {
            url: url.to_string(),
            name: "Clasp Client".to_string(),
            features: vec![
                "param".to_string(),
                "event".to_string(),
                "stream".to_string(),
            ],
            token: None,
            reconnect: true,
            reconnect_interval_ms: 5000,
            #[cfg(feature = "p2p")]
            p2p_config: None,
        }
    }

    /// Set client name
    pub fn name(mut self, name: &str) -> Self {
        self.name = name.to_string();
        self
    }

    /// Set supported features
    pub fn features(mut self, features: Vec<String>) -> Self {
        self.features = features;
        self
    }

    /// Set authentication token
    pub fn token(mut self, token: &str) -> Self {
        self.token = Some(token.to_string());
        self
    }

    /// Enable/disable auto-reconnect
    pub fn reconnect(mut self, enabled: bool) -> Self {
        self.reconnect = enabled;
        self
    }

    /// Set reconnect interval in milliseconds
    pub fn reconnect_interval(mut self, ms: u64) -> Self {
        self.reconnect_interval_ms = ms;
        self
    }

    /// Set P2P configuration (requires p2p feature)
    #[cfg(feature = "p2p")]
    pub fn p2p_config(mut self, config: clasp_core::P2PConfig) -> Self {
        self.p2p_config = Some(config);
        self
    }

    /// Build and connect
    pub async fn connect(self) -> Result<Clasp> {
        let mut client = Clasp::new(
            &self.url,
            self.name,
            self.features,
            self.token,
            self.reconnect,
            self.reconnect_interval_ms,
        );

        // Set P2P config if provided
        #[cfg(feature = "p2p")]
        {
            if let Some(p2p_config) = self.p2p_config {
                client.set_p2p_config(p2p_config);
            }
        }

        client.do_connect().await?;
        Ok(client)
    }
}
